File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    
    int n, answer;
    int[][] dungeons;
    boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        n = dungeons.length; this.dungeons = dungeons;
        
        visited = new boolean[n];
        answer = 0;
        adventure(0, k);
        return answer;
    }
    
    public void adventure(int cnt, int k) {
        answer = Math.max(answer, cnt);
        
        for (int i = 0; i < n; i++) {
            if (visited[i] || k < dungeons[i][0]) continue;
            visited[i] = true;
            adventure(cnt + 1, k - dungeons[i][1]);
            visited[i] = false;
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def adventure(k, dungeons):
    global answer
    answer = max(answer, sum(visited))

    for i in range(n):
        if visited[i] or k < dungeons[i][0]:
            continue
        visited[i] = 1
        adventure(k - dungeons[i][1], dungeons)
        visited[i] = 0
        
def solution(k, dungeons):
    global n, visited, answer
    n = len(dungeons)
    
    visited = [0] * n
    answer = 0
    adventure(k, dungeons)
    return answer